sycl/source/device_selector.cpp

```diff
@@ -43,14 +43,16 @@ int default_selector::operator()(const device &dev) const {
     const platform platform = dev.get_info<info::device::platform>();
     const std::string platformVersion =
         platform.get_info<info::platform::version>();;
+    const bool HasCudaString =
+        platformVersion.find("CUDA") != std::string::npos;
+    const bool HasOpenCLString =
+        platformVersion.find("OpenCL") != std::string::npos;
     // If using PI_CUDA, don't accept a non-CUDA device
-    if (platformVersion.find("CUDA") == std::string::npos &&
-        backend == "PI_CUDA") {
+    if (HasCudaString && HasOpenCLString && backend == "PI_CUDA") {
```

Comment:
This entire code fragment (lines 37-58) is in conflict with https://github.com/intel/llvm/pull/1522, where SYCL_BE is used as the **preferred** not as the **forced** one. For example, if the target device type is CPU, which CUDA PI plugin does not support and users specified SYCL_BE=PI_CUDA , then with this code no device will be found and an exception thrown, while in that code it will use the plugin where the CPU is supported, i.e. OpenCL. I think both semantics are useful and we should probably fork SYCL_BE into SYCL_PI_FORCE & SYCL_PI_PREFER.

Reply:
The only thing that line 51 does is that, if a platform has *both* the CUDA string and the OpenCL string, it is rejected if , and only if, the desired backend is `PI_CUDA`. This prevents the selection of the NVIDIA OpenCL platform when the user wants to use the PI CUDA backend. Everything else should work in the same way. I don't think there is a need to fork the env. variables at this point.

Reply:
Sorry, I misread the code. It is not rejecting the non-CUDA devices under SYCL_BE=PI_CUDA as I initially thought. It is still to be merged properly with the code coming in https://github.com/intel/llvm/pull/1522 (tagging @againull). Also it should be taking platform.getBackend() instead of checking platform name.

Reply:
Should I wait until #1522 and #1490 are merged? I can't use the platform.getBackend() until then.
Alternatively I can use the `is_cuda` until the other MRs are there.